### PR TITLE
feat(calling): added new calling class

### PR DESCRIPTION
**Pull request by Kesari3008**
# COMPLETES #< INSERT LINK TO ISSUE >

## This pull request addresses

Created global Calling class which will take care of below:
1. If Webex object is provided then store it inside SDK connector. Proceed with creating client objects based on the config received.
2. If Webex is not present and webexConfig is shared by developer, we will follow below:

- Initialize the Webex object, emit 'calling:ready' event once webex is ready. 
- On receiving the 'calling:ready' event, developer can trigger register() API under calling class.
- register() API will trigger WDM registration, mercury connection.
- Once above actions are completed, we will create client objects based on the calling config received.

## by making the following changes

< DESCRIBE YOUR CHANGES >

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [x] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/calling/calling/src/Calling/Calling.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+  public callSettingsClient!: ICallSettings;
+
+  public voicemailClient!: IVoicemail;
+
+  /**
+   *
+   * @param callingConfig
+   * @param webex
+   * @param webexConfig
```
Comment (meta-dipanshu-sharma) [inline packages/calling/calling/src/Calling/Calling.ts:45]: going forward we should stop adding jsdocs in ts file.

**Comment by meta-dipanshu-sharma** [comment]
I'm curious about the folder structure. Why we have `calling/calling/*` type of path?

File: packages/calling/calling/src/Calling/Calling.ts (hunk 1 above)
Comment (Kesari3008) [reply packages/calling/calling/src/Calling/Calling.ts:45]: Ohh I didn't add this..seems to have added on its own when i ran lint --fix command. Let me fix it

File: packages/webex/src/calling.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import {
```
Comment (sreenara) [inline packages/webex/src/calling.js:1]: What's this calling.js file that is created here?

File: packages/calling/calling/src/Calling/Calling.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
+/* eslint-disable tsdoc/syntax */
```
Comment (sreenara) [inline packages/calling/calling/src/Calling/Calling.ts:1]: Why is this required?

File: packages/calling/calling/src/Calling/Calling.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        SDKConnector.setWebex(this.webex);
+      }
+
+      this.initializeClients(callingConfig);
+    }
+    log.setLogger(callingConfig.logger.level);
+  }
+
+  /**
+   *
```
Comment (sreenara) [inline packages/calling/calling/src/Calling/Calling.ts:73]: Can we add a one line description above the function so that this gets added to the tsdoc automatically?

File: packages/calling/calling/src/Calling/Calling.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+   */
+  public register() {
+    const logContext = {
+      file: CALLING_FILE,
```
Comment (sreenara) [inline packages/calling/calling/src/Calling/Calling.ts:77]: This constant should be within a constants file in the Calling folder. I don't think we should import it from CallingClient.

File: packages/calling/calling/src/Calling/Calling.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      this.voicemailClient = createVoicemailClient(logger);
+    }
+  }
+
+  /**
+   * Function to return SDKConnector object.
+   *
+   * @returns SDKConnector.
+   */
+  public getSDKConnector(): ISDKConnector {
```
Comment (sreenara) [inline packages/calling/calling/src/Calling/Calling.ts:138]: This doesn't seem to be used anywhere?

File: packages/calling/calling/src/Calling/types.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import {ICallSettings} from '../CallSettings/types';
+import {CallingClientConfig, ICallingClient} from '../CallingClient/types';
+import {IContacts} from '../Contacts/types';
+import {LoggerConfig} from '../Logger/types';
+import {ISDKConnector} from '../SDKConnector/types';
+import {IVoicemail} from '../Voicemail/types';
+import {Eventing} from '../Events/impl';
+
+interface ClientConfig {
+  calling?: boolean;
```
Comment (sreenara) [inline packages/calling/calling/src/Calling/types.ts:12]: Does boolean make sense here? You're setting it as optional, which means that this field doesn't need to be present in the object. So then setting this to false doesn't make sense right?

File: packages/calling/calling/src/Calling/types.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  contact?: boolean;
+  history?: boolean;
+  settings?: boolean;
+  voicemail?: boolean;
+}
+
+export interface WebexConfig {
+  access_token: string;
```
Comment (sreenara) [inline packages/calling/calling/src/Calling/types.ts:20]: Are we not going to add any other members in the WebexConfig?
We should be giving the developer the ability to set additional parameters in the webex config right?

File: packages/calling/calling/src/Calling/types.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+export interface CallingConfig {
+  clientConfig: ClientConfig;
+  callingClientConfig: CallingClientConfig;
+  logger: LoggerConfig;
+}
+
+export interface ICalling extends Eventing<CallingEventTypes> {
+  register: () => Promise<void>;
```
Comment (sreenara) [inline packages/calling/calling/src/Calling/types.ts:30]: Please set the members first followed by the methods.

**Review by sreenara — CHANGES_REQUESTED** [review]
Please see the comments.
Also, please list out the files that are changed as part of this PR. The number of files in the PR is really large.

**Comment by Kesari3008** [comment]
New PR raised: https://github.com/webex/webex-js-sdk/pull/2908
